### PR TITLE
Make `Hash`, `Hasher` and `BuildHasher` `#[const_trait]` and make `Sip` const `Hasher`

### PR DESCRIPTION
**Pull request by onestacked**
This PR enables using Hashes in const context.

r? @fee1-dead

### CODE REVIEW

File: library/core/src/hash/sip.rs [hunk 1]
```diff
@@ -238,7 +243,11 @@ impl super::Hasher for SipHasher {
 }
 
 #[unstable(feature = "hashmap_internals", issue = "none")]
-impl super::Hasher for SipHasher13 {
+#[rustc_const_unstable(feature = "const_hash", issue = "104061")]
+impl const super::Hasher for SipHasher13
+where
+    Hasher<Sip13Rounds>: ~const super::Hasher,
+{
```
Comment (fee1-dead) [inline library/core/src/hash/sip.rs:250]: Why is this needed?

Comment (chriss0612) [reply library/core/src/hash/sip.rs]: Seems like it isn't

File: library/core/src/hash/mod.rs [hunk 2]
```diff
@@ -234,12 +235,19 @@ pub trait Hash {
...
+        //FIXME(const_iter_slice): Revert to for loop
+        //for piece in data {
+        //    piece.hash(state);
+        //}
+
+        let mut i = 0;
+        while i < data.len() {
+            data[i].hash(state);
+            i += 1;
         }
```
Comment (fee1-dead) [inline library/core/src/hash/mod.rs:251]: `data[i]` has bound checks IIRC. you could try writing a benchmark and see if its performance has changed or use `const_eval_select` to reduce runtime performance impact.

File: library/core/src/hash/sip.rs [hunk 3]
```diff
@@ -138,7 +138,7 @@ unsafe fn u8to64_le(buf: &[u8], start: usize, len: usize) -> u64 {
         out |= (unsafe { *buf.get_unchecked(start + i) } as u64) << (i * 8);
         i += 1;
     }
-    debug_assert_eq!(i, len);
+    debug_assert!(i == len);
```
Comment (fee1-dead) [inline library/core/src/hash/sip.rs:141]: We should probably revert back for the error message when #103639 has landed. So please add a fixme.

File: library/std/src/lib.rs [hunk 4]
```diff
@@ -314,6 +314,7 @@
 #![feature(maybe_uninit_uninit_array)]
 #![feature(const_maybe_uninit_uninit_array)]
 #![feature(const_waker)]
+#![feature(const_hash)]
```
Comment (fee1-dead) [inline library/std/src/lib.rs:317]: maybe move to the "only for constness" section below?

**Review by fee1-dead — CHANGES_REQUESTED** [review]
Left some nits. Also, I would like to see some tests to ensure that it works (maybe not as an ui test, but under `library/core/tests`, next to the hash tests)

File: library/core/src/hash/sip.rs (hunk 3 above)
Comment (chriss0612) [reply library/core/src/hash/sip.rs:141]: Should it be FIXME(fee1-dead) or FIXME(debug_assert_eq)?

File: library/core/src/hash/mod.rs (hunk 2 above)
Comment (chriss0612) [reply library/core/src/hash/mod.rs:251]: > Left some nits. Also, I would like to see some tests to ensure that it works (maybe not as an ui test, but under `library/core/tests`, next to the hash tests)

Should I add a new file for testing const Hash or should i make the current Tests const and add some asserts that compare the CT and RT results?

**Comment by chriss0612** [comment]
> Left some nits. Also, I would like to see some tests to ensure that it works (maybe not as an ui test, but under `library/core/tests`, next to the hash tests)

Should I add a new file or make the current tests const and add some asserts that check for differences between RT and CT?

File: library/core/src/hash/sip.rs (hunk 3 above)
Comment (fee1-dead) [reply library/core/src/hash/sip.rs:141]: `fee1-dead` is fine.

**Comment by fee1-dead** [comment]
> Should I add a new file or make the current tests const and add some asserts that check for differences between RT and CT?

You should add const tests alongside the runtime asserts. Differences between RT and CT are generally not needed to test if you copy all runtime tests and make them const alongside the original tests.

**Comment by fee1-dead** [comment]
Just the general tests. for example, in https://github.com/rust-lang/rust/blob/master/library/core/tests/hash/mod.rs you don't need `test_indirect_hasher` and below as those do not currently make sense in const contexts. You could also put the const logic in the same function as the runtime logic using const items.

File: library/core/src/hash/mod.rs (hunk 2 above)
Comment (chriss0612) [reply library/core/src/hash/mod.rs]: I used const_eval_select for now.
I hope that const `Iterator` for `Slice` won't take to long.

**Review by fee1-dead — APPROVED** [review]
